packages/playwright/src/runner/watchMode.ts

```diff
@@ -207,37 +202,32 @@ export async function runWatchModeLoop(config: FullConfigInternal): Promise<Full
         config.cliGrep = testPattern;
       else
         config.cliGrep = undefined;
-      await fsWatcher.update(config);
-      await runTests(config, failedTestIdCollector);
+      await runTests(config, testServerConnection);
       lastRun = { type: 'regular' };
       continue;
     }
 
     if (command === 'failed') {
-      config.testIdMatcher = id => failedTestIdCollector.has(id);
       const failedTestIds = new Set(failedTestIdCollector);
-      await runTests(config, failedTestIdCollector, { title: 'running failed tests' });
-      config.testIdMatcher = undefined;
+      await runTests(config, testServerConnection, { title: 'running failed tests', testIds: [...failedTestIds] });
       lastRun = { type: 'failed', failedTestIds };
       continue;
     }
 
     if (command === 'repeat') {
       if (lastRun.type === 'regular') {
-        await runTests(config, failedTestIdCollector, { title: 're-running tests' });
+        await runTests(config, testServerConnection, { title: 're-running tests' });
         continue;
       } else if (lastRun.type === 'changed') {
-        await runChangedTests(config, failedTestIdCollector, lastRun.dirtyTestFiles!, 're-running tests');
+        await runChangedTests(config, testServerConnection, lastRun.dirtyTestFiles!, 're-running tests');
       } else if (lastRun.type === 'failed') {
-        config.testIdMatcher = id => lastRun.failedTestIds!.has(id);
-        await runTests(config, failedTestIdCollector, { title: 're-running tests' });
-        config.testIdMatcher = undefined;
+        await runTests(config, testServerConnection, { title: 're-running tests', testIds: [...lastRun.failedTestIds!] });
       }
       continue;
     }
 
     if (command === 'toggle-show-browser') {
-      await toggleShowBrowser(config, originalWorkers);
+      await toggleShowBrowser();
```

Comment:
this is entirely untested, and ideally i'd like to add a test for it - but i'm not sure how to do it best. I tried writing a demo page that reads from `localStorage` across the two test runs, but that gets cleared. how could we assert that a browser is reused?

Reply:
I'd not worry about testing this piece for now.